### PR TITLE
Added basic lights (directional, point, spot)

### PR DESCRIPTION
**Pull request by jaremieromer**

### CODE REVIEW

File: source/ncengine/graphics2/diligent/ShaderFactory.h [hunk 1]
```diff
@@ -42,15 +42,18 @@ class ShaderFactory
...
     auto shaderSourceFactory = Diligent::RefCntAutoPtr<Diligent::IShaderSourceInputStreamFactory>{};
 #if NC_RUNTIME_SHADER_COMPILATION
-    engineFactory.CreateDefaultShaderSourceStreamFactory(nullptr, &shaderSourceFactory);
+    engineFactory.CreateDefaultShaderSourceStreamFactory(shadersPath.data(), &shaderSourceFactory);
     if (!shaderSourceFactory)
     {
         throw NcError{"Failed to create ShaderSourceInputStreamFactory"};
     }
+#else
+(void)shadersPath;
```
Comment (McCallisterRomer) [inline source/ncengine/graphics2/diligent/ShaderFactory.h:56]: `[[maybe_unused]]` on the param like with `engineFactory`.

File: source/ncengine/graphics2/diligent/resource/GlobalEnvironmentResource.h [hunk 2]
```diff
@@ -19,8 +20,7 @@ class GlobalEnvironmentResource
                                            Diligent::IRenderDevice& device,
                                            Diligent::IShaderResourceVariable& variable);
 
-        void Update(const CameraRenderState cameraState,
-                    Diligent::IDeviceContext& context);
+        void Update(Diligent::IDeviceContext& context, const CameraRenderState cameraState, const LightRenderState lightRenderState);
```
Comment (McCallisterRomer) [inline source/ncengine/graphics2/diligent/resource/GlobalEnvironmentResource.h]: Will you add `&` to `cameraState` and `lightRenderState`?

File: source/ncengine/graphics2/diligent/resource/base/StructuredBuffer.h [hunk 3]
```diff
@@ -50,13 +51,11 @@ class StructuredBuffer : public StructuredBufferBase
 
         explicit StructuredBuffer(Diligent::IDeviceContext& context,
                                   Diligent::IRenderDevice& device,
-                                  std::string_view name,
                                   Diligent::IShaderResourceVariable& variable,
-                                  uint32_t maxElementCount,
-                                  uint32_t initialElementCount = maxElementCount)
-            : StructuredBufferBase{name, variable, maxElementCount, initialElementCount}
+                                  StructuredBufferResourceDesc resourceDesc)
```
Comment (McCallisterRomer) [inline source/ncengine/graphics2/diligent/resource/base/StructuredBuffer.h]: `const &`

File: source/ncengine/graphics2/diligent/resource/component/ComponentResourceSignature.h [hunk 4]
```diff
@@ -23,26 +24,26 @@ class ComponentResourceSignature
                                             Diligent::IRenderDevice& device,
                                             std::string_view signatureName,
                                             uint8_t bindingIndex,
-                                            StructuredBufferResourceDesc meshRendererResourceDesc);
+                                            StructuredBufferResourceDesc meshRendererResourceDesc,
```
Comment (McCallisterRomer) [inline source/ncengine/graphics2/diligent/resource/component/ComponentResourceSignature.h]: `const &` for all

File: source/ncengine/graphics2/frontend/GraphicsFrontend.cpp [hunk 5]
```diff
@@ -9,7 +9,8 @@ auto GraphicsFrontend::BuildRenderState(ecs::Ecs world) -> FrontendRenderState
     return FrontendRenderState{
         .cameraState = m_cameraSystem.BuildState(world),
         .meshRendererState = m_meshRendererSystem.BuildState(world),
-        .materialRenderState = m_materialRegistry.BuildState()
+        .materialRenderState = m_materialRegistry.BuildState(),
+        .lightRenderState = m_lightSubsystem.BuildState(world, world, world) /** This can't be the right usage of ExplicitEcs */
```
Comment (McCallisterRomer) [inline source/ncengine/graphics2/frontend/GraphicsFrontend.cpp]: Its not. Just make one `ExplicitEcs<T, U, V>`.

File: source/ncengine/graphics2/frontend/subsystem/CameraRenderState.h [hunk 6]
```diff
@@ -7,5 +7,6 @@ namespace nc::graphics
 struct CameraRenderState
 {
     DirectX::XMMATRIX viewProjection = DirectX::XMMatrixIdentity();
+    Vector3 position = Vector3::One();
```
Comment (McCallisterRomer) [inline source/ncengine/graphics2/frontend/subsystem/CameraRenderState.h]: I highly doubt it matters, but zero seems a more sensible default for position.

File: source/ncengine/graphics2/frontend/subsystem/LightSubsystem.h [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+#include <vector>
+
+namespace nc
+{
+namespace graphics
+{
+class LightSubsystem
+{
+    public:
+        auto BuildState(ecs::ExplicitEcs<DirectionalLight, Transform> dirLightEcs,
```
Comment (McCallisterRomer) [inline source/ncengine/graphics2/frontend/subsystem/LightSubsystem.h]: Stated before, but one ecs should do.

File: source/ncengine/graphics2/frontend/subsystem/LightSubsystem.h [hunk 8]
```diff
@@ -0,0 +1,30 @@
+#pragma once
+
+#include "LightRenderState.h"
+#include "graphics2/ShaderTypes.h"
+#include "ncengine/ecs/EcsFwd.h"
+
+#include "ncengine/graphics/DirectionalLight.h"
```
Comment (McCallisterRomer) [inline source/ncengine/graphics2/frontend/subsystem/LightSubsystem.h]: You have `EcsFwd` here, so you should be able to also forward declare all component types.

File: source/ncengine/ui/editor/impl/ComponentWidgets.cpp [hunk 9]
```diff
@@ -803,13 +804,22 @@ void ParticleEmitterUIWidget(graphics::ParticleEmitter& emitter, EditorContext&,
...
+{
+    ui::InputColor3(light.color, "color");
+}
+
 void PointLightUIWidget(graphics::PointLight& light, EditorContext&, const std::any&)
 {
     constexpr auto step = 0.1f;
     constexpr auto min = 0.0f;
     constexpr auto max = 1200.0f;
+#ifndef NC_USE_DILIGENT
```
Comment (McCallisterRomer) [inline source/ncengine/ui/editor/impl/ComponentWidgets.cpp:817]: If we're only using one of the colors going forward, will you make an issue to remove it? I would, but I'm not sure what the plan is/if this is correct.

File: test/ncengine/serialize/ComponentSerialization_integration_tests.cpp [hunk 10]
```diff
@@ -155,6 +156,15 @@ TEST(ComponentSerializationTests, RoundTrip_particleEmitter_preservesValues)
     EXPECT_EQ(expectedInfo.kinematic.scaleOverTimeFactor, actualInfo.kinematic.scaleOverTimeFactor);
 }
 
+TEST(ComponentSerializationTests, RoundTrip_directionalLight_preservesValues)
+{
+    auto stream = std::stringstream{};
+    const auto expected = nc::graphics::DirectionalLight{nc::Vector3::Splat(1.0f)};
```
Comment (McCallisterRomer) [inline test/ncengine/serialize/ComponentSerialization_integration_tests.cpp]: The serialization is so trivial here that it won't matter, but I prefer making the values of 'expected's here different from their defaults to be super sure we're reading data correctly (typical deserialization workflow default constructs before overwrite).

**Review by McCallisterRomer — COMMENTED** [review]
Oops, didn't see it was a draft until now. I'll post my comments, but can rereview if you have more to do here.

File: source/ncengine/graphics2/frontend/subsystem/LightSubsystem.h (hunk 8 above)
Comment (jaremieromer) [reply source/ncengine/graphics2/frontend/subsystem/LightSubsystem.h]: Ah - I had forward declared them but then had to fully declare them when I added the vectors, but forgot to remove `EcsFwd.h`

File: source/ncengine/ui/editor/impl/ComponentWidgets.cpp (hunk 9 above)
Comment (jaremieromer) [reply source/ncengine/ui/editor/impl/ComponentWidgets.cpp:817]: Done, https://github.com/NcStudios/NcEngine/issues/807

**Review by McCallisterRomer — APPROVED** [review]

**Pull request merged**